### ISSUE TITLE
`OPTIMIZE FINAL DEDUPLICATE` query is not working after upgrading to ver. 23.7

### ISSUE THREAD

**Issue by mkaze** [ClickHouse/ClickHouse#54041]
**Describe what's wrong**

After upgrading to ver. 23.7 I am no longer able to deduplicate data using `OPTIMIZE` query. Previously, on ver. 22.8, it worked completely fine; however, after the upgrade, when I run the exact same query as before, it immediately returns without any effect (i.e. the duplicate data is still there). Here is the query I am using:

```sql
OPTIMIZE TABLE
    my_table
PARTITION
    tuple(toYYYYMM(toDate('2023-08-20')))
FINAL DEDUPLICATE BY [list_of_primary_keys...];
```

The `my_table` is a `MergeTree` table. And I am using the python client of `clickhouse` (i.e. [`clickhouse-driver`](https://github.com/mymarilyn/clickhouse-driver) ver. 0.2.2) to run the query. In clickhouse ver. 22.8, it took around 5-10 minutes to execute, and the duplicate data would be removed successfully. To make sure the python driver is not faulty, I also tested with `clickhouse-client` with no success (it immediately returns with the message: "Ok. 0 rows in set. Elapsed: 0.001 sec.").

To be precise, I am using the docker deployment of `clickhouse-server` without any custom settings/configuration.

**Comment by alexey-milovidov** [comment]
There could be a reason why it cannot do the OPTIMIZE. Check it with `SET optimize_throw_if_noop = 1` and run the query.

**Comment by mkaze** [comment]
> Check it with SET optimize_throw_if_noop = 1 and run the query.

Awesome! It turns out the issue was due to insufficient disk space:

```
DB::Exception: Cannot OPTIMIZE table: Insufficient available disk space, required 191.63 GiB. Cannot select parts for optimization.
```
Never thought of it, especially because the disk had ~80GB of free space.
Thank you @alexey-milovidov! I close the issue.
